packages/@react-spectrum/s2/src/Breadcrumbs.tsx

```diff
@@ -189,47 +291,180 @@ export interface BreadcrumbProps extends Omit<AriaBreadcrumbItemProps, 'children
+  CollectionRoot({collection}) {
+    return useCollectionRender(collection);
+  },
+  CollectionBranch({collection}) {
+    return useCollectionRender(collection);
+  }
+};
+
+let useCollectionRender = (collection: Collection<Node<unknown>>) => {
+  let {containerRef, onAction} = useContext(CollapseContext) ?? {};
+  let [visibleItems, setVisibleItems] = useState(collection.size);
+
+  let children = useMemo(() => {
+    let result: Node<any>[] = [];
+    for (let key of collection.getKeys()) {
+      result.push(collection.getItem(key)!);
+    }
+    return result;
+  }, [collection]);
+
+  let listRef = useRef<HTMLDivElement | null>(null);
+  let updateOverflow = useCallback(() => {
+    let currListRef: HTMLDivElement | null = listRef.current;
+    if (!currListRef) {
+      setVisibleItems(collection.size);
+      return;
+    }
+    let container = currListRef.parentElement;
+    if (!container) {
+      setVisibleItems(collection.size);
+      return;
+    }
+
+    let listItems = Array.from(currListRef.querySelectorAll('[data-hidden-breadcrumb]')) as HTMLLIElement[];
+    let folder = currListRef.querySelector('button') as HTMLButtonElement;
+    if (listItems.length <= 0) {
+      setVisibleItems(collection.size);
+      return;
+    }
+    let containerWidth = container.offsetWidth;
+    let containerGap = parseInt(getComputedStyle(container).gap, 10);
+    let folderGap = parseInt(getComputedStyle(folder).marginInlineStart, 10);
+    let newVisibleItems = 0;
+
+    let widths: Array<number> = [];
+    for (let breadcrumb of listItems) {
+      let width = breadcrumb.offsetWidth + 1; // offsetWidth is rounded down
+      widths.push(width);
+    }
+
+    // we know there is always at least one item because of the listItems.length check up above
+    let widthOfFirst = widths.shift()!;
+    let availableWidth = containerWidth - widthOfFirst - folderGap - folder.offsetWidth - containerGap;
+    for (let width of widths.reverse()) {
+      availableWidth -= width;
+      if (availableWidth <= 0) {
+        break;
+      }
+      availableWidth -= containerGap;
+      newVisibleItems++;
+    }
+
+    setVisibleItems(Math.max(MIN_VISIBLE_ITEMS, Math.min(MAX_VISIBLE_ITEMS, newVisibleItems)));
+  }, [collection.size, setVisibleItems]);
+
+  // making bad assumption that i can listen to containerRef when it's declared in the parent?
+  useResizeObserver({ref: containerRef, onResize: updateOverflow});
+
+  useLayoutEffect(() => {
+    if (collection.size > 0) {
+      queueMicrotask(updateOverflow);
+    }
+  }, [collection.size, updateOverflow]);
+
+  useEffect(() => {
+    // Recalculate visible tags when fonts are loaded.
+    document.fonts?.ready.then(() => updateOverflow());
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+  let sliceIndex = collection.size - visibleItems;
+
+  return (
+    <>
+      <HiddenBreadcrumbs items={children} size="M" listRef={listRef} />
```

Comment:
I'm was bit surprised that the size provided here can remain as `M` regardless of the actual size set, but it seems like size doesn't actually get used by the breadcrumb styles. Do we still need it then?

Reply:
Ah, forgot to do size changes, I'll update